pkg/koordlet/statesinformer/impl/states_pod_resources.go

```diff
@@ -0,0 +1,189 @@
+	}
+	s.nodeInformer = nodeInformer
+}
+
+func (s *podResourcesInformer) Start(stopCh <-chan struct{}) {
+	if !features.DefaultKoordletFeatureGate.Enabled(features.HookedPodResources) {
+		return
+	}
+	klog.V(2).Infof("starting pod resources informer")
+	if !cache.WaitForCacheSync(stopCh, s.nodeInformer.HasSynced) {
+		klog.Fatalf("timed out waiting for node caches to sync")
+	}
+	stub, err := newKubeletStubFromConfig(s.nodeInformer.GetNode(), s.config)
+	if err != nil {
+		klog.Fatalf("create kubelet stub, %v", err)
+	}
+	resourceClient, err := GetResourceClient("")
+	if err != nil {
+		klog.Fatalf("create resource client, %v", err)
+	}
+	s.resourceServer = &podResourcesServer{
+		kubeletStub:       stub,
+		podResourceClient: resourceClient,
+	}
+	err = s.startServer()
+	if err != nil {
+		klog.Fatalf("start grpc server, %v", err)
+	}
+}
+
+func (s *podResourcesInformer) startServer() error {
+	grpcServerFullSocketPath := filepath.Join(s.config.HookedPodResourceSocketPath, defaultKubeletSocket+".sock")
+	err := cleanup(grpcServerFullSocketPath)
+	if err != nil {
+		klog.Errorf("failed to cleanup %s: %s", grpcServerFullSocketPath, err.Error())
+		return err
+	}
+	sock, err := net.Listen("unix", grpcServerFullSocketPath)
+	if err != nil {
+		klog.Errorf("failed to listen: %s", err.Error())
+		return err
+	}
+	server := grpc.NewServer()
+	podresourcesapi.RegisterPodResourcesListerServer(server, s.resourceServer)
+	klog.Infof("Starting GRPC server, grpcServerSocketFullPath: %s", grpcServerFullSocketPath)
+	go func() {
+		err := server.Serve(sock)
+		if err != nil {
+			server.Stop()
+			klog.Fatalf("plugin exited with error %s", err.Error())
+		}
+	}()
+	return nil
+}
+
+func cleanup(grpcServerSocketFullPath string) error {
+	if err := os.Remove(grpcServerSocketFullPath); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (s *podResourcesInformer) HasSynced() bool {
+	return true
+}
+
+type podResourcesServer struct {
+	podResourceClient podresourcesapi.PodResourcesListerClient
+	kubeletStub       KubeletStub
+}
+
+func (p *podResourcesServer) List(ctx context.Context, request *podresourcesapi.ListPodResourcesRequest) (*podresourcesapi.ListPodResourcesResponse, error) {
+	response, err := p.podResourceClient.List(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	allPods, err := p.kubeletStub.GetAllPods()
+	if err != nil {
+		return nil, err
+	}
+	fillPodDevicesAllocatedByKoord(response, &allPods)
+	return response, nil
+}
+
+func fillPodDevicesAllocatedByKoord(response *podresourcesapi.ListPodResourcesResponse, allPods *corev1.PodList) {
+	deviceTypeToResourceName := map[schedulingv1alpha1.DeviceType]string{
+		schedulingv1alpha1.GPU:  string(extension.ResourceNvidiaGPU),
+		schedulingv1alpha1.RDMA: string(extension.ResourceRDMA),
+	}
+
+	for _, podResource := range response.PodResources {
+		for _, pod := range allPods.Items {
+			if pod.Namespace == podResource.Namespace && pod.Name == podResource.Name {
```

Comment:
it seems there should be a map to retrieve the items to reduce unnecessary complexity